tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/ral_legalize_to_llvm.cc

```diff
@@ -0,0 +1,556 @@
+//
+//   device: user defined string (e.g. cpu or gpu)
+//
+//   inputs_encode = type_separator.join([type_encoding for type in
+//   input_types])
+//
+//   outputs_encode = type_separator.join([type_encoding for type
+//   in output_types])
+//
+//   type_separator = '_'
+LogicalResult getDispatchOpSignatureEncoding(DispatchOp dispatch_op,
+                                             StrT& out) {
+  const char* separator = "___";
+  // append signature prefix
+  out.append(dispatch_op.call_target_name());
+  out.append(separator);
+
+  // encode backend (device) info
+  out.append(dispatch_op.backend_config());
+  out.append(separator);
+
+  // encode input types
+  Operation* op = dispatch_op.getOperation();
+  for (auto& en : llvm::enumerate(op->getOperandTypes())) {
+    if (en.index() != 0) out.append("_");
+    if (failed(getTypeEncoding(op->getContext(), en.value(), out)))
+      return failure();
+  }
+  out.append(separator);
+
+  // encode output types
+  for (auto& en : llvm::enumerate(op->getResultTypes())) {
+    if (en.index() != 0) out.append("_");
+    if (failed(getTypeEncoding(op->getContext(), en.value(), out)))
+      return failure();
+  }
+  if (!op->getNumResults()) out.append("void");
+  return success();
+}
+
+// Loads a global op at the current insertion point and returns the loaded
+// value.
+Value loadGlobalString(OpBuilder& builder, const Location& loc,
+                       GlobalOp globalOp) {
+  MLIRContext* ctx = builder.getContext();
+  Value globalPtr = builder.create<LLVM::AddressOfOp>(loc, globalOp);
+  Value cst0 = builder.create<LLVM::ConstantOp>(
+      loc, IntegerType::get(ctx, 64),
+      builder.getIntegerAttr(builder.getIndexType(), 0));
+  return builder.create<LLVM::GEPOp>(
+      loc, LLVM::LLVMPointerType::get(IntegerType::get(ctx, 8)), globalPtr,
+      ValueRange{cst0, cst0});
+}
+
+// Creates a global const string op named `name` using the value if not exists
+// and returns the Loaded value of this global op.
+Value loadOrCreateGlobalString(PatternRewriter& rewriter, Operation* op,
+                               StringRef name, StringRef value) {
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+  GlobalOp globalOp = module.lookupSymbol<GlobalOp>(name);
+  if (!globalOp) {
+    OpBuilder::InsertionGuard guard(rewriter);
+    // Double-check under the lock
+    if (globalOp = module.lookupSymbol<GlobalOp>(name)) {
+      return loadGlobalString(rewriter, op->getLoc(), globalOp);
+    }
+    OpBuilder::InsertPoint ip = rewriter.saveInsertionPoint();
+    rewriter.setInsertionPointToStart(module.getBody());
+
+    auto type = LLVM::LLVMArrayType::get(IntegerType::get(op->getContext(), 8),
+                                         value.size());
+    globalOp = rewriter.create<LLVM::GlobalOp>(
+        op->getLoc(), type, /*isConstant=*/true, LLVM::Linkage::Internal, name,
+        rewriter.getStringAttr(value), /*alignment=*/0);
+
+    rewriter.restoreInsertionPoint(ip);
+  }
```

Comment:
Should we check that if there is an existing globalOp it also matches the provided `value`?

Reply:
Thanks! I'll add a assert here.